### PR TITLE
Use ASN1_TIME for handling time comparisons in ca.c:do_updatedb

### LINKED ISSUE #13944: updatedb expires certificates with expiration date having 4-digit years before 2100.
In a private CA, I have some very long running certificates. In index.txt one of the expiration dates is 20730104163541Z meaning the certificate is valid until early 2073. If I run an updatedb command on this CA, this certificate is marked as "expired". I assume this is, because only the first two digits are interpreted as a year and this certificate therefore looks as if it would have been expired in 2020. I'm using openssl 1.1.1d.

### PR DESCRIPTION
**Pull request by arminfuerst**
Fixes #13944
   + changed ASN1_UTCTIME to ASN1_TIME
   + removed all Y2K code from do_updatedb
   + changed compare to ASN1_TIME_compare

### CODE REVIEW

File: apps/ca.c [hunk 1]
```diff
@@ -2223,62 +2223,48 @@ static int get_certificate_status(const char *serial, CA_DB *db)
...
 
-            if (db_y2k == a_y2k) {
-                /* all on the same y2k side */
-                if (strcmp(rrow[DB_exp_date], a_tm_s) <= 0) {
-                    rrow[DB_type][0] = DB_TYPE_EXP;
-                    rrow[DB_type][1] = '\0';
-                    cnt++;
+            exp_date = ASN1_TIME_new();
+            if (exp_date == NULL)
+                return -1;
```
Comment (t8m) [inline apps/ca.c:2249]: You're leaking a_tm here.

File: apps/ca.c [hunk 2]
```diff
@@ -2223,62 +2223,48 @@ static int get_certificate_status(const char *serial, CA_DB *db)
...
 
-                    BIO_printf(bio_err, "%s=Expired\n", rrow[DB_serial]);
-                }
-            } else if (db_y2k < a_y2k) {
+            if (!ASN1_TIME_set_string(exp_date, rrow[DB_exp_date])) {
+                ASN1_TIME_free(exp_date);
+                return -1;
```
Comment (t8m) [inline apps/ca.c:2253]: And here as well.

**Review by t8m — COMMENTED** [review]
Have you tested the code? Unfortunately I do not think we have a test of the -updatedb option of the ca application. It would be nice to have one in the test_ca recipe, unfortunately it would make the PR not cherry-pickable from 1.1.1 to master. You could submit it as a separate PR against master if you wanted though.

**Comment by arminfuerst** [comment]
I made manual tests:
- The certificate that I mentioned in the bug is handled properly
- Certificates in the near future are handled properly
- Certificates in the near past are handled properly
I do not have any certificates available having expired long ago, especially not before 2000-01-01.
I also checked whether there is a testing infrastructure available to add tests, but I didn't find any - at least not in the 1.1.1 branch.

**Comment by t8m** [comment]
This is sufficient for me.

**Comment by t8m** [comment]
Please fix the possible leaks and I approve.

**Comment by arminfuerst** [comment]
Ok, I'll have a look at the leaks - can't promise how fast.
About the testing: I there a testing infrastructure in the master branch? Having tests is certainly a good idea but I doubt I could build up an infrastructure being consistent to the rest of openssl.

**Comment by t8m** [comment]
> Ok, I'll have a look at the leaks - can't promise how fast.
> About the testing: I there a testing infrastructure in the master branch? Having tests is certainly a good idea but I doubt I could build up an infrastructure being consistent to the rest of openssl.

In master there are already various tests for the ca application in `test/recipes/80-test_ca.t`. However there is no test involving the `-updatedb` option.

**Review by t8m — APPROVED** [review]
LGTM now!

**Comment by arminfuerst** [comment]
> > Ok, I'll have a look at the leaks - can't promise how fast.
> > About the testing: I there a testing infrastructure in the master branch? Having tests is certainly a good idea but I doubt I could build up an infrastructure being consistent to the rest of openssl.
> 
> In master there are already various tests for the ca application in `test/recipes/80-test_ca.t`. However there is no test involving the `-updatedb` option.

I had a look at this infrastructure and I see a major blocker to write tests for it:
I would need to modify "now". My idea is to:

1. Create a very long running CA in the past (I would create one before Y2K).
2. Create some certificates with relevant test values (expiration < Y2K, expiration > Y2K, expiration 2049, expiration >= 2050).
3. Run updatedb on dates in the times between two expiration dates. I would make a whole testrun for each time gap (to ensure also the correct expiration for the certificate expired < Y2K when the first updatedb is calles > 2050) as well as a series of updatedbs.

Any ideas how I could realize this? I never wrote tests in C where I had to override the system time, perhaps there is a simple way I don't know?

**Comment by richsalz** [comment]
The `ca` command takes `-startdate` and `-enddate` flags, will that not be good enough?

**Comment by arminfuerst** [comment]
> The `ca` command takes `-startdate` and `-enddate` flags, will that not be good enough?

I don't think so, because I need to manipulate the system time when running the `updatedb` command. Do `startdate` and `enddate` allow values in the past?

**Comment by richsalz** [comment]
> I don't think so, because I need to manipulate the system time when running the updatedb command. Do startdate and enddate allow values in the past?

Yes.  Note that this is only for certs, not other commands.

**Comment by arminfuerst** [comment]
> > I don't think so, because I need to manipulate the system time when running the updatedb command. Do startdate and enddate allow values in the past?
> 
> Yes. Note that this is only for certs, not other commands.

So one solution could be to introduce a parameter to specify the date `updatedb` should use instead of the current time. And since `-startdate` and `-enddate` are not working for the CA itself, this would also have to be introduced. Would this be an acceptable solution or is this change to deep?
BTW: Is this PR the best place for this discussion?

**Comment by t8m** [comment]
I do not think it is worth it. The parameter would be useful only for testing.

**Review by paulidale — APPROVED** [review]

**Comment by arminfuerst** [comment]
> I do not think it is worth it. The parameter would be useful only for testing.

I was thinking about different approaches. One idea was to call the CA commands with fake system times. In Linux (and probably most supported Unix-like systems), this could be achieved relatively easy using the "faketime" command. I have no idea (and it's close to impossible for me to test that) whether it's possible to compile this for Windows.
The second issue coming to my mind is, that this would introduce an external dependency for building openssl.
On the other hand: having automated tests for code at least running on some platforms is already an advantage to having none at all.
Alternatively we could move the logic into a library, and modify the existing ca.c to contain as little logic as possible and create an additional command (ca_test.c) where we could add code like fake times we need for tests.
Any ideas about these thoughts?

**Comment by t8m** [comment]
> Alternatively we could move the logic into a library, and modify the existing ca.c to contain as little logic as possible and create an additional command (ca_test.c) where we could add code like fake times we need for tests.

With 3.0 this is doable. The ca_test could even live in the test subdirectory and link that library (well, it does not have to be a library, just refactored out c file that would contain the common parts of the code between the test and the ca application). Look at how cmp_client_test is built in the test/build.info

I am going to merge this PR. If you want to add the test in master, you're welcome. Please open a new PR for that.

**Comment by t8m** [comment]
Merged to master and 1.1.1 branch. Thank you for the contribution!

**Comment by arminfuerst** [comment]
> > Alternatively we could move the logic into a library, and modify the existing ca.c to contain as little logic as possible and create an additional command (ca_test.c) where we could add code like fake times we need for tests.
> 
> With 3.0 this is doable. The ca_test could even live in the test subdirectory and link that library (well, it does not have to be a library, just refactored out c file that would contain the common parts of the code between the test and the ca application). Look at how cmp_client_test is built in the test/build.info
> 
> I am going to merge this PR. If you want to add the test in master, you're welcome. Please open a new PR for that.

This was my plan anyway. If I really make relevant progress there, is there a way to get feedback for PoC or support for specific questions (best practices in openssl) without starting a PR? I dislike the idea of opening a PR when knowing in advance it is far from being really productive...

**Comment by t8m** [comment]
PRs are quite cheap and are the best way to review even the PoC. Do not worry about opening it. You can also always force push a completely new code to the PoC PR branch if you need to start from scratch. You do not have to create a new branch and new PR in that case.
If you have just a question, no real code to commit, please open an issue of the question type.

**Pull request closed without merging**